lib/cn/kvset_split.c

```diff
@@ -0,0 +1,652 @@
+
+        rc1 = keycmp(split_key, split_klen, kblk->kb_koff_min, kblk->kb_klen_min);
+        if (rc1 < 0) {
+            return k;
+        } else {
+            int rc2 = keycmp(split_key, split_klen, kblk->kb_koff_max, kblk->kb_klen_max);
+            if (rc2 <= 0) {
+                *overlap = (rc2 < 0);
+                return k;
+            }
+        }
+    }
+
+    return ks->ks_st.kst_kblks;
+}
+
+static merr_t
+kblocks_split(
+    struct kvset            *ks,
+    const struct key_obj    *split_kobj,
+    struct kvset_split_work *work,
+    struct kvset_split_res  *result)
+{
+    struct hlog *hlog_left = work[LEFT].hlog;
+    struct hlog *hlog_right = work[RIGHT].hlog;
+    struct kvset_mblocks *blks_left = &result->blks[LEFT];
+    struct kvset_mblocks *blks_right = &result->blks[RIGHT];
+    bool overlap = false;
+    uint32_t split_idx;
+    uint8_t *hlog;
+    merr_t err;
+
+    split_idx = get_kblk_split_index(ks, split_kobj, &overlap);
+    assert(split_idx <= ks->ks_st.kst_kblks);
+
+    /* Add kblocks in [0, split_idx - 1] to the left kvset
+     * Also generate a composite hlog for the left kvset and the right kvset
+     */
+    for (uint32_t i = 0; i < split_idx; i++) {
+        err = blk_list_append(&blks_left->kblks, ks->ks_kblks[i].kb_kblk.bk_blkid);
+        if (err)
+            goto errout;
+
+        kbr_read_hlog(&ks->ks_kblks[i].kb_kblk_desc, &hlog);
+        hlog_union(hlog_left, hlog);
+    }
+
+    if (overlap) {
+        struct kvset_kblk *kblk = &ks->ks_kblks[split_idx];
+        struct kblock_desc kbd;
+        struct blk_list kblks[2];
+        merr_t err;
+
+        kbd.cn = cn_tree_get_cn(ks->ks_tree);
+        kbd.kd_mbd = &kblk->kb_kblk_desc;
+        kbd.kd_wbd = &kblk->kb_wbt_desc;
+
+        for (int i = LEFT; i <= RIGHT; i++)
+            blk_list_init(&kblks[i]);
+
+        /* split kblock at split_idx */
+        err = kblock_split(&kbd, split_kobj, kblks);
+        if (err)
+            goto errout;
+        assert(kblks[LEFT].n_blks > 0 && kblks[RIGHT].n_blks > 0);
+
+        /* Append kblks[LEFT] to the left kvset, kblks[RIGHT] to the right kvset, and both
+         * kblks[LEFT] and kblks[RIGHT] to the commit list
+         */
+        for (uint32_t i = 0; i < kblks[LEFT].n_blks; i++) {
+            uint64_t blkid = kblks[LEFT].blks[i].bk_blkid;
+
+            err = blk_list_append(&blks_left->kblks, blkid);
+            if (!err)
+                err = blk_list_append(&result->blks_commit, blkid);
+
+            if (err)
+                goto errout;
+        }
+
+        for (uint32_t i = 0; i < kblks[RIGHT].n_blks; i++) {
+            uint64_t blkid = kblks[RIGHT].blks[i].bk_blkid;
+
+            err = blk_list_append(&blks_right->kblks, blkid);
+            if (!err)
+                err = blk_list_append(&result->blks_commit, blkid);
+
+            if (err)
+                goto errout;
+        }
+
+        /* TODO: Would it be accurate to use the left and right kblock's hlog here?
```

Comment:
Are you going to fix this before merge?

Reply:
Nope. I need to understand the consequence of this, so most likely in a follow-on PR.